From 1dd808948873f42fbe224f313dcd5ba0ddae58cf Mon Sep 17 00:00:00 2001
From: Brian Ryner <bryner@google.com>
Date: Tue, 4 May 2021 14:27:37 +1000
Subject: [PATCH 3/3] Support cross-compiling to Mac ARM64.

Incorporates a previous patch to allow 'readelf' to be the host
readelf tool.
---
 configure.ac | 18 +++++++++++++++---
 setup.py     |  2 +-
 2 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/configure.ac b/configure.ac
index 5ec781affe..9eea1af301 100644
--- a/configure.ac
+++ b/configure.ac
@@ -395,12 +395,19 @@ then
 	*-*-vxworks*)
 	    ac_sys_system=VxWorks
 	    ;;
+	*-*-darwin*)
+	    ac_sys_system=Darwin
+	    ;;
 	*)
 		# for now, limit cross builds to known configurations
 		MACHDEP="unknown"
 		AC_MSG_ERROR([cross build not supported for $host])
 	esac
-	ac_sys_release=
+	if test "$ac_sys_system" = "Darwin"; then
+	    ac_sys_release=`uname -r`  # Needed for some tests below.
+	else
+	    ac_sys_release=
+	fi
     else
 	ac_sys_system=`uname -s`
 	if test "$ac_sys_system" = "AIX" \
@@ -444,6 +451,11 @@ if test "$cross_compiling" = yes; then
 	*-*-vxworks*)
 		_host_cpu=$host_cpu
 		;;
+	*-*-darwin*)
+		# The rest of the build system does not expect
+		# _PYTHON_HOST_PLATFORM to include the host cpu on Mac.
+		_host_cpu=
+		;;
 	*)
 		# for now, limit cross builds to known configurations
 		MACHDEP="unknown"
@@ -1198,9 +1210,9 @@ then
 fi
 
 AC_CHECK_TOOLS([READELF], [readelf], [:])
-if test "$cross_compiling" = yes; then
+if test "$cross_compiling" = yes -a "$ac_sys_system" != "Darwin"; then
     case "$READELF" in
-	readelf|:)
+	:)
 	AC_MSG_ERROR([readelf for the host is required for cross builds])
 	;;
     esac
diff --git a/setup.py b/setup.py
index 0a2a435c60..7fb486fd87 100644
--- a/setup.py
+++ b/setup.py
@@ -871,7 +871,7 @@ class PyBuildExt(build_ext):
             os.makedirs(self.build_temp)
         # Determine if readline is already linked against curses or tinfo.
         if do_readline:
-            if CROSS_COMPILING:
+            if CROSS_COMPILING and not MACOS:
                 ret = os.system("%s -d %s | grep '(NEEDED)' > %s" \
                                 % (sysconfig.get_config_var('READELF'),
                                    do_readline, tmpfile))
-- 
2.31.1.527.g47e6f16901-goog

